### PR TITLE
Add PBFT architecture intro and network overview

### PR DESCRIPTION
**Pull request by chenette**
- Add intro text to summarize Sawtooth PBFT architecture at a high level

- Add "Network Overview" section to describe fault tolerance (max number of
  faulty nodes), view changes, sequence numbers, and info stored by each node.
  Also, summarize on-chain settings with a link to that section.

- Delete the last section, "Node Information Storage", because that information
  is now in "Network Overview"

Signed-off-by: Anne Chenette <chenette@bitwise.io>

### CODE REVIEW

File: docs/source/architecture.rst [hunk 1]
```diff
@@ -1,17 +1,142 @@
...
+time.
+
+For example, a four-node network (the minimum required number of nodes)
+can tolerate one faulty node. Increasing the size of the network reduces the
+likelihood that that all :math:`\frac{n - 1}{3}` nodes would be faulty at the
+same time.
+
+Network `liveness
+<https://en.wikipedia.org/wiki/Liveness#Liveness_and_safety>`__
+is guaranteed as long as there are less than :math:`2f+1` unreachable nodes,
```
Comment (ltseeley) [inline docs/source/architecture.rst]: I don't think this is true; can you explain where this information is from?

File: docs/source/architecture.rst [hunk 2]
```diff
@@ -1,17 +1,142 @@
...
+
+The Sawtooth PBFT algorithm is a voting-based consensus algorithm with Byzantine
+fault tolerance, which ensures `safety and liveness
+<https://en.wikipedia.org/wiki/Liveness#Liveness_and_safety>`__.
+The network can tolerate a certain number of "bad" nodes. As long as this number
+is not exceeded, the network will work properly. In addition, blocks committed
+by nodes are final, so there are no forks in the network.
+
+All nodes on the network participate in the consensus process by verifying
+candidate blocks and exchanging consensus messages to communicate their progress
```
Comment (ltseeley) [inline docs/source/architecture.rst]: Requested some changes to wording similar to this in the previous PR; can you make similar changes here? Let me know if you have any questions about specifics.

File: docs/source/architecture.rst [hunk 3]
```diff
@@ -1,17 +1,142 @@
...
+
+.. _network-overview-label:
+
+Network Overview
+================
+
+Fault Tolerance
+---------------
+
+A PBFT network consists of nodes that are ordered from 0 to `n-1`, where
```
Comment (ltseeley) [inline docs/source/architecture.rst]: Might be useful here to mention that the ordering is determined by the list in the on-chain setting (mentioned later, but I think it's useful to call it out early for those that might wondering how the order is determined here).

File: docs/source/architecture.rst [hunk 4]
```diff
@@ -1,17 +1,142 @@
...
+  to the validator.
+
+* The secondary nodes monitor the health of the primary and request a change
+  if the primary appears out of order or malicious.
+
+A `view` is the period of time that a given node is the primary. The next
+primary is selected in a round-robin (circular) fashion, according to the order
+of nodes listed in the ``sawtooth.consensus.pbft.peers``
+:doc:`on-chain setting <on-chain-setting>`.
+In a four-node network, for example, node 0 is the primary at view 0,
```
Comment (ltseeley) [inline docs/source/architecture.rst]: Would be helpful to specify, parenthetically, which node is node 0, node 1, etc. I.e. the 0th node in the list, the 1st node in the list, etc.

File: docs/source/architecture.rst [hunk 5]
```diff
@@ -1,17 +1,142 @@
...
+counting both node and network failures. That is, at least third of the nodes
+(rounded up) must be reachable at all times.
+
+View Changes: Choosing a New Primary
+------------------------------------
+
+A PBFT network has one leader, called the `primary node`. All other nodes on the
+network are called `secondary nodes`.
+
+* The primary requests new blocks from the validator, notifies the network when
```
Comment (ltseeley) [inline docs/source/architecture.rst]: I would simplify the requests new blocks/notifies the network bit to just saying that the primary builds and publishes new blocks.

File: docs/source/architecture.rst [hunk 6]
```diff
@@ -1,17 +1,142 @@
...
+  a new block is available, and reports consensus results (commit or fail) back
```
Comment (ltseeley) [inline docs/source/architecture.rst]: While reporting the results to the validator is technically true, all nodes do this with their own validator, so I don't think it's helpful to mention that here.

File: docs/source/architecture.rst [hunk 7]
```diff
@@ -1,17 +1,142 @@
...
+  to the validator.
+
+* The secondary nodes monitor the health of the primary and request a change
```
Comment (ltseeley) [inline docs/source/architecture.rst]: The most important role of the secondaries is to vote on blocks, which I think is worth mentioning here. Also, "monitor the health of the primary" might be a bit misreading, since that may suggest they something more than what they actually do. All they're doing is checking if the primary is malicious (sent a bad message) or faulty (doesn't do it's job in time), or if the network somehow gets stuck (doesn't make progress for a while). It's really kind of "dumb" in that it's not doing particularly intelligent "monitoring".

File: docs/source/architecture.rst [hunk 8]
```diff
@@ -1,17 +1,142 @@
...
+at view 7, the formula `7 mod 4` determines that node 3 is the primary.
+
+The Sawtooth PBFT algorithm changes the primary at regular intervals, as well as
+when the secondary nodes determine that the current primary is faulty.
+See :ref:`view-changing-mode-label` for a description of this process.
+
+Sequence Numbers
+----------------
+
+The primary assigns a `sequence number` for each new block (called a "candidate
```
Comment (ltseeley) [inline docs/source/architecture.rst]: The primary doesn't really *assign* a sequence number in Sawtooth PBFT (unlike traditional PBFT), because the sequence number is already defined as the block number.

File: docs/source/architecture.rst [hunk 9]
```diff
@@ -1,17 +1,142 @@
...
+
+A PBFT network consists of nodes that are ordered from 0 to `n-1`, where
+`n` is the total number of nodes in the network.
+
+The PBFT algorithm guarantees network `safety
+<https://en.wikipedia.org/wiki/Liveness#Liveness_and_safety>`__
+as long as the number of faulty nodes remains below the required percentage.
+The maximum number of faulty nodes that the network can tolerate is determined
+by the formula :math:`f = \frac{n - 1}{3}`. In other words, no more than one
+third of the nodes (rounded down) can be out of order or dishonest at any given
```
[the remark this answers is not in the document]
Comment (chenette) [reply docs/source/architecture.rst]: @agunde406 or @ltseeley : How about "unreachable"?

File: docs/source/architecture.rst (hunk 1 above)
Comment (chenette) [reply docs/source/architecture.rst]: @ltseeley - It's adapted from this source: http://ug93tad.github.io/pbft/  

I was trying to find a way to work in the "2f +1" formula from an earlier draft of the PBFT architecture document.

File: docs/source/architecture.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (agunde406) [reply docs/source/architecture.rst]: I'm not sure what you are trying to say here?

[the remark this answers is not in the document]
Comment (chenette) [reply docs/source/architecture.rst]: (However, I will delete this paragraph per ltseeley's comment.)

File: docs/source/architecture.rst (hunk 5 above)
Comment (chenette) [reply docs/source/architecture.rst]: I was trying to avoid saying that the primary node publishes the new blocks.  My understanding is that the primary tells the validator that it's okay for the validator to publish the block.

Comment (chenette) [reply docs/source/architecture.rst]: Or does the consensus engine actually handle the mechanics of publishing a block?

File: docs/source/architecture.rst (hunk 8 above)
Comment (chenette) [reply docs/source/architecture.rst]: I think this sentence (which was in the original doc) is referring to the general PBFT algorithm. I propose rewording it thusly:

The PBFT algorithm specifies that each block has a `sequence number`. Sawtooth PBFT uses the block number as the sequence number.

File: docs/source/architecture.rst (hunk 7 above)
Comment (chenette) [reply docs/source/architecture.rst]: I'll change this to:

"The secondary nodes vote on the new blocks and send messages if the primary   
  appears unreachable or malicious."

File: docs/source/architecture.rst (hunk 3 above)
Comment (chenette) [reply docs/source/architecture.rst]: Done.

File: docs/source/architecture.rst [hunk 10]
```diff
@@ -1,17 +1,142 @@
...
+time.
+
+For example, a four-node network (the minimum required number of nodes)
+can tolerate one faulty node. Increasing the size of the network reduces the
+likelihood that that all :math:`\frac{n - 1}{3}` nodes would be faulty at the
```
[the remark this answers is not in the document]
Comment (chenette) [reply docs/source/architecture.rst]: Done

File: docs/source/architecture.rst (hunk 2 above)
Comment (chenette) [reply docs/source/architecture.rst]: Done

File: docs/source/architecture.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (chenette) [reply docs/source/architecture.rst]: The "In other words" sentence is explaining the formula in words.  I changed "out of order" to "unreachable" (in the sense that either the node isn't working or there's a network problem).

File: docs/source/architecture.rst (hunk 6 above)
Comment (chenette) [reply docs/source/architecture.rst]: Hmmm.  I'm looking for a short phrase to summarize what the primary does in the "finalize" phase without implying that the PBFT algorithm is responsible for actually committing the block. Do you have any suggestions?

File: docs/source/architecture.rst (hunk 5 above)
Comment (chenette) [reply docs/source/architecture.rst]: After chatting, I agree that we don't need to make this distinction at the beginning of the section. (It's described in more detail later.)

File: docs/source/architecture.rst [hunk 11]
```diff
@@ -1,17 +1,134 @@
...
+by nodes are final, so there are no forks in the network.
+
+The nodes on the network send many messages to reach consensus, commit blocks,
+and maintain a healthy leader node, called a `primary node`. The network
+switches to a new primary (called a `view change`) at regular intervals, as well
+as when the current primary is faulty.
+
+* The primary node builds and publishes blocks.
+
+* The other nodes (called `secondary nodes`) vote on the new blocks and send
```
Comment (ltseeley) [inline docs/source/architecture.rst]: I would reword this to just be "vote on blocks" instead of "vote on the new blocks". Also, I think "send messages if the primary appears unreachable or malicious" is too vague in that "send messages" doesn't really tell you anything about what they're doing if the primary is unreachable or malicious (they send messages for just about everything). Not totally sure what the best way to word this is, though.

File: docs/source/architecture.rst [hunk 12]
```diff
@@ -1,17 +1,134 @@
...
+
+In a four-node network, for example, the first node (node 0) is the primary at
+view 0, the second node (node 1) is the primary at view 1, and so on.  When the
+network gets to view 4, it will "wrap back around" so that node 0 is the primary
+again.
+
+The algorithm uses the formula `p = v mod n` to determine the next
+primary. In this formula, `p` is the primary, `v` is the view number, and `n` is
+the total number of nodes in the network. For example, if a four-node network is
+at view 7, the formula `7 mod 4` determines that node 3 (the fourth node) is the
```
Comment (ltseeley) [inline docs/source/architecture.rst]: I think calling out that node 3 is the fourth node here could be a little confusing and is unnecessary.

File: docs/source/architecture.rst [hunk 13]
```diff
@@ -1,17 +1,134 @@
...
+See :ref:`view-changing-mode-label` for a description of this process.
+
+Sequence Numbers
+----------------
+
+With PBFT, each new block has a `sequence number`. Sawtooth PBFT uses the block
+number as the sequence number.
+
+For example, a node that is on sequence 10 is evaluating block 10.
+Note that this is actually the 11th block, because both sequence and block
```
Comment (ltseeley) [inline docs/source/architecture.rst]: I'd remove this sentence; we never really refer to a block as the "11th" block, we'd just refer to it as block 10.

File: docs/source/architecture.rst [hunk 14]
```diff
@@ -1,17 +1,134 @@
...
+primary.
+
+The Sawtooth PBFT algorithm changes the primary at regular intervals, as well as
+when the secondary nodes determine that the current primary is faulty.
+See :ref:`view-changing-mode-label` for a description of this process.
+
+Sequence Numbers
+----------------
+
+With PBFT, each new block has a `sequence number`. Sawtooth PBFT uses the block
```
Comment (ltseeley) [inline docs/source/architecture.rst]: Blocks don't actually have sequence numbers. The notion of a "sequence number" is really only relevant to the PBFT nodes themselves; i.e. you would say that the node is on sequence number 10, and that it is attempting to commit block 10 at sequence number 10.

File: docs/source/architecture.rst [hunk 15]
```diff
@@ -276,11 +393,10 @@ View changing mode has the following steps:
    operation.
 
 The next primary node is determined by the node ID, in sequential order, based
-based on the order of nodes in the ``sawtooth.consensus.pbft.peers`` on-chain
-setting. The first node on the list has ID 0 and is the first primary for view
-0; the next node has ID 1 and is the second primary for view 1; and so on.
-After all nodes on the list have been the primary, the algorithm returns to
-node 0.
+on the order of nodes in the ``sawtooth.consensus.pbft.peers`` on-chain setting.
+The first node on the list has ID 0 and is the first primary for view 0; the
```
Comment (ltseeley) [inline docs/source/architecture.rst]: "ID 0", "ID 1", etc. is outdated information; this was from back in the beginning where nodes had numbered ID's. Now, the only concept of a node ID is the node's public key. This paragraph should be updated to reflect that and match what's in the "View Changes: Choosing a New Primary" section

File: docs/source/architecture.rst (hunk 11 above)
Comment (chenette) [reply docs/source/architecture.rst]: How about "The other nodes ... vote on new blocks and the health of the leader." ?

File: docs/source/architecture.rst (hunk 14 above)
Comment (chenette) [reply docs/source/architecture.rst]: I'll change it to use similar wording to the PBFT blog post.

File: docs/source/architecture.rst (hunk 15 above)
Comment (chenette) [reply docs/source/architecture.rst]: In order to avoid duplicated text, I'll replace this info with a link to the (correct) info in "View Changes: Choosing a New Primary".

File: docs/source/architecture.rst [hunk 16]
```diff
@@ -1,17 +1,136 @@
...
+The Sawtooth PBFT algorithm changes the primary at regular intervals, as well as
+when the secondary nodes determine that the current primary is faulty.
+See :ref:`view-changing-mode-label` for a description of this process.
+
+Sequence Numbers
+----------------
+
+In addition to moving through a series of views, the network moves through a
+series of `sequence numbers`. In Sawtooth PBFT, a node's sequence number is
+the same as block number of the next block in the chain. For example, a node
```
Comment (ltseeley) [inline docs/source/architecture.rst]: Missing "the": "same as block number"

File: docs/source/architecture.rst [hunk 17]
```diff
@@ -1,17 +1,136 @@
...
+that is on sequence 10 has already committed block 9 and is evaluating block 10.
```
Comment (ltseeley) [inline docs/source/architecture.rst]: I'd say "sequence number 10", not "sequence 10"
